## Pandas – Basics 01

![](img/pandas.svg)

In this lesson, we're going to introduce some of the basics of [Pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/overview.html), a powerful Python library for working with tabular data like CSV files.

We will cover how to:

* Import Pandas
* Read in a CSV file
* Explore and filter data
* Make simple plots and data visualizations
* Write a CSV file

---

### Setup

#### Import pandas

In [23]:
import pandas as pd

The above `import` statement not only imports the Pandas library but also gives it an alias or nickname — `pd`. This alias will save us from having to type out the entire words `pandas` each time we need to use it. Many Python libraries have commonly used aliases like `pd`.

#### Set Display Settings

By default, Pandas will display 60 rows and 20 columns. I often change [Pandas' default display settings](https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html) to show more rows or columns.

In [ ]:
pd.options.display.max_rows = 100

#### Read in CSV File

To read in a CSV file, we will use the function `pd.read_csv()` and insert the name of our desired file path. 

In [ ]:
bellevue_df = pd.read_csv('./datasets/data/bellevue_almshouse.csv', delimiter=",")

This creates a Pandas [DataFrame object](https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html#dataframe) — often abbreviated as *df*, e.g., *bellevue_df*. A DataFrame looks and acts a lot like a spreadsheet. But it has special powers and functions that we will discuss in the next few lessons.

When reading in the CSV file, we also specified the `encoding` and `delimiter`. The `delimiter` specifies the character that separates or "delimits" the columns in our dataset. For CSV files, the delimiter will most often be a comma. (CSV is short for *Comma Separated Values*.) Sometimes, however, the delimiter of a CSV file might be a tab (`\t`) or, more rarely, another character.


---

### Display Data


#### Plain

We can display a DataFrame in a Jupyter notebook simply by running a cell with the variable name of the DataFrame.

In [ ]:
bellevue_df

There are a few important things to note about the DataFrame displayed here:

* Index
    * The ascending numbers in the very left-hand column of the DataFrame is called the Pandas *Index*. You can select rows based on the Index.
    * By default, the Index is a sequence of numbers starting with zero. However, you can change the Index to something else, such as one of the columns in your dataset.

* Truncation
    * The DataFrame is truncated, signaled by the ellipses in the middle `...` of every column.
    * The DataFrame is truncated because we set our default display settings to 100 rows. Anything more than 100 rows will be truncated. To display all the rows, we would need to alter Pandas' default display settings yet again.

* Rows x Columns
    * Pandas reports how many rows and columns are in this dataset at the bottom of the output (9584 x 8 columns).

#### Display First *n* Rows

To look at the first *n* rows in a DataFrame, we can use a method called `.head()`.

In [ ]:
bellevue_df.head(2)

In [ ]:
bellevue_df.head(10)

#### Display Random Sample

To look at a random sample of rows, we can use the `.sample()` method.

In [ ]:
bellevue_df.sample(10)

---

### Explore the dataframe

#### Simple

To get important info about all the columns in the DataFrame, we can use the `.info()` method.

In [ ]:
bellevue_df.info()

#### Shape

To explicitly check for how many rows vs columns make up a dataset, we can use the `.shape` property.

In [ ]:
bellevue_df.shape

There are 9584 rows and 8 columns.

#### Columns

We can also check the column names of the DataFrame with [`.columns`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.columns.html) property.

In [ ]:
bellevue_df.columns

#### Data Types

It's important to always check the data types in your DataFrame. For example, sometimes numeric values will accidentally be interpreted as a string object. To perform calculations on this data, you would need to first convert that column from a string to an integer.

In [ ]:
bellevue_df.dtypes

Just like Python has different data types, Pandas has different data types, too. These data types are automatically assigned to columns when we read in a CSV file. We can check these Pandas data types with the [`.dtypes` method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html).



| **Pandas Data Type** |  **Explanation**                                                                                   |
|:-------------:|:---------------------------------------------------------------------------------------------------:|
| `object`         | string                                                                               |
| `float64`         | float                                               |
| `int64`       | integer                                                        |
| `datetime64`       |  date time              

#### Calculate Summary Statistics

To calculate summary statistics for every column in our DataFrame, we can use the `.describe()` method.

In [ ]:
bellevue_df.describe()

By default, `.describe()` will only compute columns with numerical data. To include all columns, we can use `include='all'`.

In [21]:
bellevue_df.describe(include='all')

,date_in,first_name,last_name,age,disease,profession,gender,children
count,9584,9580,9584,9534.000000,6497,8565,9584,37
unique,653,523,3142,NaN,75,172,5,36
top,1847-05-24,Mary,Kelly,NaN,sickness,laborer,m,Child
freq,113,979,137,NaN,2706,3108,4958,2
mean,NaN,NaN,NaN,30.332604,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,14.179608,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,0.080000,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,21.000000,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,28.000000,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,39.000000,NaN,NaN,NaN,NaN


Here are some insights that can be gleaned from these summary statistics:
- For the column **date_in**, the earliest recorded admission to the Bellevue Almshouse (**first**) is 1847-05-24 
- For the column **first_name**, the most frequently occurring first name (**top**) is Mary, which appears 979 times (**freq**)
- For the column **last_name**, the most frequently occurring last name (**top**) is Kelly, which appears 137 times (**freq**)
- For the column **age**, average age in the dataset (**mean**) is 30, the youngest (**min**) is .8, and the oldest (**max**) is 97
- For the columns **disease** and **profession**, there are 75 unique (**unique**) diseases and 172 unique (**unique**) professions
- For the column **children**, there are 37 rows that include information about children (**count**) (i.e., rows that do not have blank values)

---

### Working with the DataFrame

#### Select Columns

To select a column from the DataFrame, we will type the name of the DataFrame followed by square brackets and a column name in quotations marks.

In [ ]:
bellevue_df['disease']

Technically, a single column in a DataFrame is a [*Series* object](https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html#dsintro).

In [ ]:
type(bellevue_df['disease'])

A Pandas Series is a one-dimensional array of labeled data. Think of it as a list with index labels.

*Key Characteristics*

A Series has two main components:

- Values - the actual data (numbers, strings, dates, etc.)
- Index - labels for each value (like row numbers or custom names)

*Create a Series*

`ages = pd.Series([25, 30, 35, 40], index=['Alice', 'Bob', 'Charlie', 'Diana'])`

Result:

`Alice      25`

`Bob        30`

`Charlie    35`

`Diana      40`

*Series vs DataFrame*
- Series = single column (1D)
- DataFrame = multiple columns (2D, like a full spreadsheet)

A Series object displays differently than a DataFrame object. To select a column as a DataFrame and not as a Series object, we will use two square brackets.

In [ ]:
bellevue_df[['disease']]

In [ ]:
type(bellevue_df[['disease']])

By using two square brackets, we can also select multiple columns at the same time.

In [ ]:
bellevue_df[['first_name', 'last_name', 'disease']]

Heads up! The code below will cause an error.

See what happens if we try to select multiple columns as a Series...

In [ ]:
bellevue_df['first_name', 'last_name', 'disease']

#### Count Values

To count the number of *unique* values in a column, we can use the `.value_counts()` method.

Some were diagnosed with medically recognizable illnesses, including “fever,” “dropsy” and “neuralgia.” Others were diagnosed with “diseases” that made visible the ways in which immigrants were failing to meet the expectations of urban citizenship. These included “destitution” and “recent emigrant.” Neither of these diagnoses reflected an immigrant’s health. Nevertheless, they were treated as pathologies, and those pathologies governed city officials perceptions of immigrants. Sickness, injuries or destitution were subsumed under the pathology of “recent emigrant.” This diagnosis also determined immigrants’ paths through the New York City public health system.

In [ ]:
bellevue_df['disease'].value_counts()

Look through the so-called "diseases" recorded in the Bellevue Almshouse data and consider what these categories reflect about New York City in the 1840s, particularly with regard to immigration status.

To select the top 10 most frequent values in the "disease" column, we can combine `value_counts()` with regular Python list slicing.

In [ ]:
bellevue_df['disease'].value_counts()[:10]

In [ ]:
bellevue_df['profession'].value_counts()[:10]

In a similar vein, consider what these "professions" reflect about New York City in the 1840s.

#### Make and Save Plots

Pandas makes it easy to create plots and data visualizations. We can make a simple plot by adding `.plot()` to any DataFrame or Series object that has appropriate numeric data.

In [ ]:
import matplotlib.pyplot as plt

bellevue_df['disease'].value_counts()[:10].plot(kind='bar', title='Bellevue Almshouse:\nMost Frequent "Diseases"')

 We specify the title with the `title=` parameter and the kind of plot by altering the `kind=` parameter:
* ‘bar’ or ‘barh’ for bar plots

* ‘hist’ for histogram

* ‘box’ for boxplot

* ‘kde’ or ‘density’ for density plots

* ‘area’ for area plots

* ‘scatter’ for scatter plots

* ‘hexbin’ for hexagonal bin plots

* ‘pie’ for pie plots
 
For example, to make a horizontal bar chart, we can set `kind='barh'`

In [ ]:
bellevue_df['disease'].value_counts()[:10].plot(kind='barh',title='Bellevue Almshouse:\nMost Frequent "Diseases"').get_figure().savefig('plots/Bellevue_diseases_barh.png')

To make a pie chart, we can set `kind='pie'`

In [ ]:
bellevue_df['profession'].value_counts()[:10].plot(kind='pie', figsize=(10, 10), title='Bellevue Almshouse:\nMost Frequent "Professions"').get_figure().savefig('plots/Bellevue_professions_pie.png')

To save a plot as an image file or PDF file, we can assign the plot to a variable called `ax`, short for axes.

Then we can use `ax.figure.savefig('FILE-NAME.png')` or `ax.figure.savefig('FILE-NAME.pdf')`.

In [ ]:
ax = bellevue_df['profession'].value_counts()[:10].plot(kind='pie', figsize=(10, 10), title='Bellevue Almshouse:\nMost Frequent "Professions"')
ax.figure.savefig('plots/Bellevue-professions_pie.pdf')

If your plot is being cut off in the image, see "Prevent Labels From Getting Cut Off".

#### Filter/Subset Data

We can filter a Pandas DataFrame to select only certain values. Filtering data by certain values is similar to selecting columns.

We type the name of the DataFrame followed by square brackets and then, instead of inserting a column name, we insert a True/False condition. For example, to select only rows that contain the value "teacher," we insert the condition `bellevue_df['profession'] == 'teacher'`

In [ ]:
bellevue_df[bellevue_df['profession'] == 'teacher']

It can be helpful to isolate this condition and see that it produces a long list of True/False pairs for every row.

In [ ]:
bellevue_df['profession'] == 'teacher'

Filtering DataFrames can sometimes get confusing and unwieldy (as conditions within conditions pile up like Russian dolls). It can be helpful to make a separate variable for a filter, as below.

In [ ]:
teacher_filter = bellevue_df['profession'] == 'teacher'

In [ ]:
bellevue_df[teacher_filter]

In a similar vein, it's often useful to make a new variable for a filtered DataFrame. For example, let's say we wanted to look at only the women in the dataset and see the most commons professions.

In [ ]:
women_filter = bellevue_df['gender'] == 'w'
bellevue_women = bellevue_df[women_filter]
bellevue_women['profession'].value_counts()

In [ ]:
ax = bellevue_women['profession'].value_counts()[:4].plot(kind='pie', figsize=(10,10), title='Bellevue Almshouse:\nMost Frequent "Professions" Among Women')
ax.figure.savefig('plots/Bellevue-women-professions_pie.png')

There's a lot we can do with filters beyond exact value matches with an equals operator `==`.

We can also incorporate `>`, `<`, `>=`, `<=` with integers, floats, and even dates. For example, we can filter the DataFrame for only people who arrived to the Bellevue Almshouse on or after '1847-04-17'

In [ ]:
date_filter = bellevue_df['date_in'] >= '1847-04-17'
bellevue_df[date_filter]

#### Write to CSV

To output a new CSV file, we can use the `.to_csv` method with a name for the file in quotation marks. 

Here's how we might output a new CSV file that only includes rows with women.

In [ ]:
bellevue_women.to_csv("csv/Bellevue_women.csv", encoding='utf-8', index=False)

In addition to a filename, we're also specifying that the encoding is utf-8 and that the Index (the bolded left-most column) is not included in the CSV file.